### PR TITLE
Parser: basic function declaration

### PR DESCRIPTION
**Pull request by nitnelave**
This change is [Reviewable](https://reviewable.io/reviews/nitnelave/hopper/23)

### CODE REVIEW

File: src/ast/base_types.h [hunk 1]
```diff
@@ -4,19 +4,31 @@ namespace ast {
...
 class Identifier {
  public:
-  explicit Identifier(std::string name) : name_(std::move(name)) {}
+  explicit Identifier(std::string name, lexer::Range location,
+                      bool is_uppercase, bool absolute = false)
+      : name_(std::move(name)),
+        location_(std::move(location)),
+        absolute_(absolute),
+        is_uppercase_(is_uppercase) {}
   const std::string& to_string() const { return qualified_name(); }
```
Comment (Brolock) [inline src/ast/base_types.h:13]: Shouldn't to_string() be virtual, otherwise why does it exists?

Comment (nitnelave) [reply src/ast/base_types.h:13]: Better to move the review to reviewable.io, but this is just to provide a common interface to all printable types.

**Comment by Brolock** [comment]
Review status: 0 of 42 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/lexer/token.h, line 245 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-KorW1A7MQYfnuYSSssh:-KorW1A7MQYfnuYSSssi:b-jm2m90) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/src/lexer/token.h#L245)):*
> ```objc
>     if (value().is<std::string>()) return text();
>     if (value().is<int64_t>()) return std::to_string(int_value());
>     assert(false);
> ```

That's not a nice assert message

---

*[src/lexer/token.h, line 250 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-KorX55xBSejIT86v6Lo:-KorX55yUN48AU1kLT-5:bxd04og) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/src/lexer/token.h#L250)):*
> ```objc
>   std::string to_string() const {
>     std::stringstream ss;
>     ss << "{" << type();
> ```

Shouldn't it be something like type().to_string() ? Since type is an int I don't know if you want to print "{9} at location" in case of empty string?

---

*[src/parser/parser.cc, line 40 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-Kor_hRHvqBn_HuZNZMZ:-Kor_hRIp_kVzCO284IT:b-knamjl) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/src/parser/parser.cc#L40)):*
> ErrorOr<ast::Identifier> Parser::parse_type_identifier(bool simple) {

> ErrorOr<ast::Identifier> Parser::parse_value_identifier(bool simple) {

Have the same body except for the ParseError returned, maybe having a 3rd function called by both to avoid code duplication

---

*[src/parser/parser.cc, line 60 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-KordQ3FIhVYyzWjiP7Y:-KordQ3FIhVYyzWjiP7Z:b2nhpzb) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/src/parser/parser.cc#L60)):*
> ```cpp
>     } else {
>       return Identifier(text.str(), location.range(), true, absolute);
>       ;
> ```

Uh?

---

*[src/parser/parser.cc, line 66 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-KordSqF1ydllDC_BugV:-Kord04B7TAdvMTtz_qN:bdb5wmp) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/src/parser/parser.cc#L66)):*
> ```cpp
>     RETURN_IF_ERROR(get_token());
>     return Identifier(text.str(), location.range(), false, absolute);
>     ;
> ```

I guess there is something with those ;
Do you mind to enlighten me?

---

*[src/parser/parser.cc, line 115 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-KorhDN8SS31VdaXLX82:-KorhDN8SS31VdaXLX83:b-bdpmjp) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/src/parser/parser.cc#L115)):*
> ```cpp
>   assert(current_token().type() == TokenType::VAL ||
>          current_token().type() == TokenType::MUT);
>   bool mut = current_token().type() == TokenType::MUT;
> ```

mut declared here and only used on return statement 

---

*[src/parser/scoped_location.cc, line 10 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-Korof7fi1kl7ItEXYyy:-Korof7fi1kl7ItEXYyz:b-sfvtbk) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/src/parser/scoped_location.cc#L10)):*
> ```cpp
>     : parser_(parser), position_(parser_->current_token().location().begin) {}
> 
> lexer::Range ScopedLocation::range() const {
> ```

Why would you want to return a non-const Range? For what I saw you only look at the range and never change it

---

*[test/test_utils/lexing.cc, line 12 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-KorqaD4JC3WRwG-XvZi:-KorqaD4JC3WRwG-XvZj:bnem69g) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/test/test_utils/lexing.cc#L12)):*
> ```cpp
>     result.emplace_back(std::move(tok));
>   }
>   return std::move(result);
> ```

Isn't RVO supposed to do that instead of you?

---

*[tools/lcov.sh, line 9 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-Korqqo5P5uE1BlUA2JT:-Korqqo5P5uE1BlUA2JU:b2aoa2l) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/tools/lcov.sh#L9)):*
> ```Shell
> lcov --gcov-tool "${DIRNAME}/llvm-gcov" \
>   --rc lcov_branch_coverage=1 \
>   --rc lcov_excl_line="(^[-0-9: #]*};?$)|(LCOV_EXCL_LINE)|(assert\(false)" \
> ```

I don't know what that is but I see assert(false) I comment

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nitnelave/hopper/23)*

**Comment by nitnelave** [comment]
Review status: 0 of 42 files reviewed at latest revision, 9 unresolved discussions.

---

*[src/lexer/token.h, line 245 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-KorW1A7MQYfnuYSSssh:-KosUW_GOOei-6kbzFS_:b-2rcr03) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/src/lexer/token.h#L245)):*
<details><summary><i>Previously, Brolock (Marcelin Dupraz) wrote…</i></summary><blockquote>

That's not a nice assert message
</blockquote></details>

Told it to straighten up and be nicer.

---

*[src/lexer/token.h, line 250 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-KorX55xBSejIT86v6Lo:-KosUilRdbyOgtjAsE-T:birf5ho) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/src/lexer/token.h#L250)):*
<details><summary><i>Previously, Brolock (Marcelin Dupraz) wrote…</i></summary><blockquote>

Shouldn't it be something like type().to_string() ? Since type is an int I don't know if you want to print "{9} at location" in case of empty string?
</blockquote></details>

Operator << is overloaded for TokenType.

---

*[src/parser/parser.cc, line 40 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-Kor_hRHvqBn_HuZNZMZ:-KosUvbGSZ088RkOLopW:bq08sqe) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/src/parser/parser.cc#L40)):*
<details><summary><i>Previously, Brolock (Marcelin Dupraz) wrote…</i></summary><blockquote>

> ErrorOr<ast::Identifier> Parser::parse_type_identifier(bool simple) {

> ErrorOr<ast::Identifier> Parser::parse_value_identifier(bool simple) {

Have the same body except for the ParseError returned, maybe having a 3rd function called by both to avoid code duplication

</blockquote></details>

There's a slight difference: a boolean is negated in the condition. I already introduced parser_identifier for that purpose, but made it usable in more places. I think the code isn't worth factoring further with another function that would have a weird xor in the middle.

---

*[src/parser/parser.cc, line 60 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-KordQ3FIhVYyzWjiP7Y:-KosVH8wzE9hixJhgvJB:b-ir4r1p) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/src/parser/parser.cc#L60)):*
<details><summary><i>Previously, Brolock (Marcelin Dupraz) wrote…</i></summary><blockquote>

Uh?
</blockquote></details>

What, you don't like spaces? :p Fixed

---

*[src/parser/parser.cc, line 66 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-KordSqF1ydllDC_BugV:-KosVMMA6BcBf0EpCpxG:b-7wklqq) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/src/parser/parser.cc#L66)):*
<details><summary><i>Previously, Brolock (Marcelin Dupraz) wrote…</i></summary><blockquote>

I guess there is something with those ;
Do you mind to enlighten me?
</blockquote></details>

All praise the copy pasta!

---

*[src/parser/parser.cc, line 115 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-KorhDN8SS31VdaXLX82:-KosVQ-WX7XZ45nZPgYh:b4jmhql) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/src/parser/parser.cc#L115)):*
<details><summary><i>Previously, Brolock (Marcelin Dupraz) wrote…</i></summary><blockquote>

mut declared here and only used on return statement 
</blockquote></details>

Yes, but the value of current_token() changes.

---

*[src/parser/scoped_location.cc, line 10 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-Korof7fi1kl7ItEXYyy:-KosVUhr21qZjGP7oEbj:blmuxr4) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/src/parser/scoped_location.cc#L10)):*
<details><summary><i>Previously, Brolock (Marcelin Dupraz) wrote…</i></summary><blockquote>

Why would you want to return a non-const Range? For what I saw you only look at the range and never change it
</blockquote></details>

I create a range, if they want to change it, good for them, it's theirs (cf std::move, for instance).
In other words, why would I want it const?

---

*[test/test_utils/lexing.cc, line 12 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-KorqaD4JC3WRwG-XvZi:-KosVhfdXuN0J1IdJo3Z:b-iurn7o) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/test/test_utils/lexing.cc#L12)):*
<details><summary><i>Previously, Brolock (Marcelin Dupraz) wrote…</i></summary><blockquote>

Isn't RVO supposed to do that instead of you?
</blockquote></details>

Not in that case. Since we don't have a copy constructor, the code doesn't compile before being optimized into an RVO

---

*[tools/lcov.sh, line 9 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-Korqqo5P5uE1BlUA2JT:-KosVqRF7ssHOrGHKzUu:b-u9wah8) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/tools/lcov.sh#L9)):*
<details><summary><i>Previously, Brolock (Marcelin Dupraz) wrote…</i></summary><blockquote>

I don't know what that is but I see assert(false) I comment
</blockquote></details>

exclude assert(false) lines from coverage.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nitnelave/hopper/23)*

**Comment by nitnelave** [comment]
Review status: 0 of 46 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/lexer/lexer.cc, line 53 at r5](https://reviewable.io:443/reviews/nitnelave/hopper/23#-KonbSKUXc_p-Stb1xuG-r5-53:-KoshRrPeH5TH8eIWIM7:b-896fix) ([raw file](https://github.com/nitnelave/hopper/blob/9830f45b7488f41afe2d078d6ec1c9c0452f1117/src/lexer/lexer.cc#L53)):*
<details><summary><i>Previously, codacy-bot (Codacy Bot) wrote…</i></summary><blockquote>

Codacy Issue found: [Class 'LexerHelper' has a constructor with 1 argument that is not explicit.](https://www.codacy.com/app/nitnelave1/hopper/pullRequest?prid=766374)
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nitnelave/hopper/23)*

**Comment by Brolock** [comment]
:lgtm:

---

Review status: 0 of 46 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/parser/scoped_location.cc, line 10 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-Korof7fi1kl7ItEXYyy:-KoufUINuBk0zms5ockI:b3jbvse) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/src/parser/scoped_location.cc#L10)):*
> ood for them, it's theirs (cf std::move, for instance).
> In other words, why would I want it const?

I just don't like giving non const things to people when it's not supposed to change, but yeah who cares

---

*[test/test_utils/lexing.cc, line 12 at r2](https://reviewable.io:443/reviews/nitnelave/hopper/23#-KorqaD4JC3WRwG-XvZi:-Koufr43UsrAHiFdfGS6:brqasjt) ([raw file](https://github.com/nitnelave/hopper/blob/21ccb1131b30d0fd4ca0edb24ce8554251a4192d/test/test_utils/lexing.cc#L12)):*
<details><summary><i>Previously, nitnelave (nitnelave) wrote…</i></summary><blockquote>

Not in that case. Since we don't have a copy constructor, the code doesn't compile before being optimized into an RVO
</blockquote></details>

Oh, forgot about it

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/nitnelave/hopper/23#-:-Kov_CPIZ64VQSWJhQc9:bnfp4nl)*

**Pull request merged**